framework/src/modules/chain/modules/blocks/chain.js

```diff
@@ -567,48 +567,32 @@ __private.loadSecondLastBlockStep = function(secondLastBlockId, tx) {
 };
 
 /**
- * Reverts changes on confirmed columns of mem_account for one transaction
- * @param {Object} transaction - transaction to undo
- * @param {Object} oldLastBlock - secondLastBlock
+ * Reverts confirmed transactions due to block deletion
+ * @param {Object} block - secondLastBlock
  * @param {Object} tx - database transaction
  */
-__private.undoConfirmedStep = function(transaction, oldLastBlock, tx) {
-	return new Promise((resolve, reject) => {
-		// Retrieve sender by public key
-		modules.accounts.getAccount(
-			{ publicKey: transaction.senderPublicKey },
-			(accountErr, sender) => {
-				if (accountErr) {
-					// Fatal error, memory tables will be inconsistent
-					library.logger.error(
-						'Failed to get account for undoing transaction to confirmed state',
-						accountErr
-					);
-					return setImmediate(reject, accountErr);
-				}
-				// Undoing confirmed transaction - refresh confirmed balance (see: logic.transaction.undo, logic.transfer.undo)
-				// WARNING: DB_WRITE
-				return modules.transactions.undoConfirmed(
-					transaction,
-					oldLastBlock,
-					sender,
-					undoConfirmedErr => {
-						if (undoConfirmedErr) {
-							// Fatal error, memory tables will be inconsistent
-							library.logger.error(
-								'Failed to undo transaction to confirmed state of account',
-								undoConfirmedErr
-							);
-							return setImmediate(reject, undoConfirmedErr);
-						}
-						return setImmediate(resolve);
-					},
-					tx
-				);
-			},
-			tx
-		);
-	});
+__private.undoConfirmedStep = async function(block, tx) {
+	if (block.transactions.length <= 0) {
+		return;
+	}
+
+	const {
+		stateStore,
+		transactionResponses,
+	} = await modules.processTransactions.undoTransactions(
+		block.transactions,
+		tx
+	);
+
+	const unappliedTransactionResponse = transactionResponses.find(
```

Comment:
Why is the error throwing logic here (in `__private.undoConfirmedStep`) instead of in `processTransactions.undoTransactions`?

Reply:
Because `processTransactions.undoTransactions` is going to be used in the transaction pool as well. And in transaction pool, we don't want the transaction responses to effectively decide which transactions to remove from the queue and which transactions to move to a different queue.

Reply:
I think that's the way process transactions was designed `applyTransactions` doesn't throw either and is left for the user to handle the error based on the state of `transactionResponses`

Reply:
Ok.